heRho
=========

Dependencies (via [conda](https://conda.io/miniconda.html))
-------
```
conda create -n herho && \
conda activate herho && \
conda install pandas numpy pybedtools tqdm docopt scikit-allel tabix ### Add jupyter + ML dependencies
```

Workflow
-------

[0] Preprocessing: tally pairwise variant configurations using variant data or simulations
-------
A. Variant data - heRho/heRho_tally_pairwise_counts_vcf.py

Input: VCF file assuming one diploid per read group. Optional BED file delegating regions to tally.

Output: Data tsv compliant with step [1] (example at heRho/example_data.tsv)

Usage
-------

```
heRho_tally_pairwise_counts_vcf.py

Usage: 
 heRho_tally_pairwise_counts_vcf.py -v <FILE> [-h -b <FILE> -d <INT> -t <INT> -s <STR> -c <STR> -f <STR>]

Options:
 -v, --vcf <FILE>                       VCF file
 -b, --bed <FILE>                       Bed file
 -d, --distance <INT>                   Max pairwise distance (default 1000)
 -t, --threads <INT>                    Thread limit (parallelised per chromosome, 1 by default)
 -s, --samples <STR>                    Comma separated list of samples to analyse (all by default)
 -c, --chromosomes <STR>                Comma separated list of chromosomes to analyse (all by default)
 -f, --file_prefix <STR>                Optional file prefix

```

B. Simulations - 
```
```

[1] Estimating parameters
-------

A. A fully tutorialised workflow given an input file generated by [0] can be run using the jupyter notebook 'heRho/heRho_full_tutorial.ipynb'.

B. A stand-alone script to estimate recombination rates per chromosome can be run using  'heRho/heRho_stand_alone.py'.
